# Sol Golomb’s Rectangle Puzzle

This problem by Solomon Golomb was presented by Gary Antonik in his 14/4/14 New York Times [Numberplay column](http://wordplay.blogs.nytimes.com/2014/04/14/rectangle):

><i>Say you’re given the following challenge: create a set of five rectangles that have sides of length 1, 2, 3, 4, 5, 6, 7, 8, 9 and 10 units. You can combine sides in a variety of ways: for example, you could create a set of rectangles with dimensions 1 x 3, 2 x 4, 5 x 7, 6 x 8 and 9 x 10.
>
>1. How many different sets of five rectangles are possible?
>
>2. What are the maximum and minimum values for the total areas of the five rectangles?
>
>3. What other values for the total areas of the five rectangles are possible?
>
>4. Which sets of rectangles may be assembled to form a square?</i>

To me, these are interesting questions because, first, I have a (slight) personal connection to Solomon Golomb (my former colleague at USC) and to Nelson Blachman (the father of my colleague Nancy Blachman), who presented the problem to  Antonik, and second, I find it interesting that the problems span the range from mathematical to computational. Let's answer them.

## 1. How many different sets of five rectangles are possible?

This is a basic [combinatorics](http://en.wikipedia.org/wiki/Combinatorics) or counting problem. I will present *three* methods to count the sets. If all goes well they will give the same answer. The example set of rectangles given in the problem was

> {1 &times; 3, 2 &times; 4, 5 &times; 7, 6 &times; 8, 9 &times; 10}
    
and in general it would be

> {A &times; B, C &times; D, E &times; F, G &times; H, I &times; J}

The question is: how many distinct ways can we assign the integers 1 through 10 to the variables A through J?
    
**Method 1: Count all permutations and divide by repetitions:** There are 10 variables to be filled, so there are 10! = 3,628,800 permutations.  But if we fill the first two variables with 1 &times; 3, that is the same rectangle as 3 &times; 1. So divide 10! by 2<sup>5</sup> to account for the fact that each of 5 rectangles can appear 2 ways.  Similarly, if we fill A and B with 1 &times; 3, that yields the same set as if we filled C and D with 1 &times; 3.  So divide again by 5! (the number of permutations of 5 things) to account for this.
That gives us:

> <a href="https://www.google.com/search?ie=UTF-8#q=10!+%2F+2%5E5+/+5!">10! / 2<sup>5</sup> / 5!</a> = 945.  

(It is always a relief when this "count and divide" method comes out to a whole number.)

**Method 2: Count without repetitions**: in each rectangle of the example set the smaller component is listed first, and in each set, the rectangles with smaller first components are listed first.  An alternate to "count and divide" is to count directly how many sets there are that respect this ordering.  We'll work from left to right.  How many choices are there for variable A?  Only one: A must always be 1, because we agreed that the smallest number comes first. Then, given A, there are 9 remaining choices for B.  For C, given A and B, there is again only one choice: C must be the smallest of the remaining 8 numbers (it will be 3 if the first rectangle was 1 &times; 2; otherwise it will be 2, but either way there is only one choice).  That leaves 7 choices for D, 5 for F, 3 for H and 1 for J. So:

> 9 &times; 7 &times; 5 &times; 3 &times; 1 = 945.

(It is always a relief when two methods give the same answer.)
          
**Method 3: Write a program to enumerate the sets:** We'll represent the 1 &times; 3 rectangle as the tuple `(1, 3)` and the example set of rectangles as the set

    {(1, 3), (2, 4), (5, 7), (6, 8), (9, 10)}

We'll write a program to generate all possible sets of rectangles, following method 2, and then just count how many there are. To implement method 2, the minimum side will always be the first element, A, in an (A, B) pair. We iterate through all possible values for B, and then join that pair with all possible rectangles made from the remaining sides. We also have to handle the case when there are no sides; then there is one possible set of rectangles: the empty set.

In [1]:
def rectangle_sets(sides):
    "Given a set of sides, list all distinct sets of rectangles that can be made."
    if not sides:
        return [ set() ]
    else:
        A = min(sides)
        return [ {(A, B)} | other_rects
                for B in sides if B is not A
                for other_rects in rectangle_sets(sides - {A, B}) ]

In [2]:
rectangle_sets({1, 2, 3, 4})

[{(1, 2), (3, 4)}, {(1, 3), (2, 4)}, {(1, 4), (2, 3)}]

We see above that there are three possible rectangle sets for 4 sides (the 1 can pair with any other side, and then a second rectangle can only be made one way from the remaining 2 sides). Below we count how many can be made with 10 sides:

In [3]:
sides = {1, 2, 3, 4, 5, 6, 7, 8, 9, 10}

all_sets = rectangle_sets(sides)

len(all_sets)

945

(It is a relief that once again we get the same answer, 945.)  

I don't want to print all 945 sets, but let's peek at every 100th one:

In [4]:
all_sets[::100]

[{(1, 2), (3, 4), (5, 6), (7, 8), (9, 10)},
 {(1, 2), (3, 10), (4, 8), (5, 6), (7, 9)},
 {(1, 3), (2, 10), (4, 6), (5, 7), (8, 9)},
 {(1, 4), (2, 10), (3, 5), (6, 8), (7, 9)},
 {(1, 5), (2, 9), (3, 8), (4, 6), (7, 10)},
 {(1, 6), (2, 9), (3, 5), (4, 7), (8, 10)},
 {(1, 7), (2, 9), (3, 4), (5, 8), (6, 10)},
 {(1, 8), (2, 7), (3, 9), (4, 5), (6, 10)},
 {(1, 9), (2, 7), (3, 5), (4, 6), (8, 10)},
 {(1, 10), (2, 7), (3, 4), (5, 8), (6, 9)}]

## 2. What are the maximum and minimum values for the total areas of the five rectangles?

I think I know this one, but I'm not completely sure.  I know that a rectangle with a fixed perimeter has maximum area when it is a square. My guess is that the maximum total area occurs when each rectangle is *almost* square: a 9 &times; 10 rectangle; 7 &times; 8; and so on.  So that would give us a maximum area of:

In [5]:
(1 * 2) + (3 * 4) + (5 * 6) + (7 * 8) + (9  * 10)

190

And the minimum area should be when the rectangles deviate the most from squares: a 1 &times; 10, and so on:

In [6]:
(1 * 10) + (2 * 9) + (3 * 8) + (4 * 7) + (5 * 6)

110

Since I am not sure, I will double check by finding the rectangle sets with the min and max total areas:

In [7]:
def total_area(rectangles): return sum(w * h for (w, h) in rectangles)

In [8]:
max(all_sets, key=total_area)

{(1, 2), (3, 4), (5, 6), (7, 8), (9, 10)}

In [9]:
min(all_sets, key=total_area)

{(1, 10), (2, 9), (3, 8), (4, 7), (5, 6)}

This verifies my intuitions. But I still don't completely understand the situation.  Suppose there are *N* sides that are not  consecutive integers.  Will the maximum total area always be formed by combining the two biggest sides, and so on?

## 3. What other values for the total areas of the five rectangles are possible?

I have no idea how to figure this out mathematically from first principles, but it is easy to compute with the code we already have:

In [10]:
areas = {total_area(s) for s in all_sets}
print(sorted(areas))

[110, 111, 112, 113, 114, 115, 116, 117, 118, 119, 120, 121, 122, 123, 124, 125, 126, 127, 128, 129, 130, 131, 132, 133, 134, 135, 136, 137, 138, 139, 140, 141, 142, 143, 144, 145, 146, 147, 148, 149, 150, 151, 152, 153, 154, 155, 156, 157, 158, 159, 160, 161, 162, 163, 164, 165, 166, 167, 168, 169, 170, 171, 172, 173, 174, 175, 177, 178, 179, 180, 181, 182, 183, 184, 186, 187, 190]


Is there a more succint way to describe these values? Let's see ...

In [11]:
set(range(110, 191)) - areas

{176, 185, 188, 189}

The answer is: "All the integers between 110 and 190 inclusive, except 176, 185, 188, and 189."

## 4. Which sets of rectangles may be assembled to form a square?

The only way I can think about this is to write a program; I don't see any way to work it out by hand.  I do know that the total area will have to be a perfect square, and since the total area is between 110 and 191, that means either 121, 144, or 169:

In [12]:
perfect_squares = {i ** 2 for i in range(100)}

sorted((total_area(s), s) 
       for s in all_sets if total_area(s) in perfect_squares)

[(121, {(1, 6), (2, 9), (3, 10), (4, 8), (5, 7)}),
 (121, {(1, 6), (2, 10), (3, 8), (4, 9), (5, 7)}),
 (121, {(1, 6), (2, 10), (3, 9), (4, 7), (5, 8)}),
 (121, {(1, 7), (2, 10), (3, 6), (4, 9), (5, 8)}),
 (121, {(1, 8), (2, 6), (3, 10), (4, 9), (5, 7)}),
 (121, {(1, 8), (2, 7), (3, 10), (4, 6), (5, 9)}),
 (121, {(1, 8), (2, 9), (3, 7), (4, 6), (5, 10)}),
 (121, {(1, 8), (2, 10), (3, 5), (4, 9), (6, 7)}),
 (121, {(1, 9), (2, 7), (3, 6), (4, 10), (5, 8)}),
 (121, {(1, 9), (2, 7), (3, 8), (4, 6), (5, 10)}),
 (121, {(1, 9), (2, 7), (3, 10), (4, 5), (6, 8)}),
 (121, {(1, 9), (2, 8), (3, 6), (4, 7), (5, 10)}),
 (121, {(1, 10), (2, 5), (3, 9), (4, 8), (6, 7)}),
 (121, {(1, 10), (2, 8), (3, 7), (4, 5), (6, 9)}),
 (144, {(1, 3), (2, 9), (4, 10), (5, 7), (6, 8)}),
 (144, {(1, 3), (2, 10), (4, 7), (5, 9), (6, 8)}),
 (144, {(1, 3), (2, 10), (4, 8), (5, 7), (6, 9)}),
 (144, {(1, 5), (2, 6), (3, 8), (4, 10), (7, 9)}),
 (144, {(1, 7), (2, 4), (3, 8), (5, 9), (6, 10)}),
 (144, {(1, 7), (2, 9), (3, 5),

In [13]:
len(_)

35

So 35 out of 945 rectangle sets *might* be assembled into a square; we don't know yet if they *can* be. 

# Packing Rectangles into a Square

I would like to see *how* the rectangles are packed into the square, not just *which* sets can be packed, so I'll need a data structure to hold information on which rectangle goes where, and a way to visually display the results.
I'll represent a *Grid* as a two-dimensional array: a list of rows, each of which is a list of cells, with the idea that each cell will be covered by a rectangle by assigning `grid[y][x] = rect` (and the special rectangle `empty` will be used to mark a cell as uncovered). I only need to deal with square grids:

In [14]:
empty = (0, 0)

def Square(n):
    "Create an [n x n] grid of empty cells."
    return [[empty for col in range(n)] 
            for row in range(n)]

In [15]:
Square(5)

[[(0, 0), (0, 0), (0, 0), (0, 0), (0, 0)],
 [(0, 0), (0, 0), (0, 0), (0, 0), (0, 0)],
 [(0, 0), (0, 0), (0, 0), (0, 0), (0, 0)],
 [(0, 0), (0, 0), (0, 0), (0, 0), (0, 0)],
 [(0, 0), (0, 0), (0, 0), (0, 0), (0, 0)]]

Let's make a function to place a single rectangle onto a grid:  `place_rectangle_at((w, h), grid, (x0, y0))` places a rectangle of width *w* and height *h* onto a grid at position (x<sub>0</sub>, y<sub>0</sub>).  If the rectangle overlaps a non-empty cell, or goes off the grid, we return `None` to indicate that this is not a legal placement. If the rectangle does fit, we return a new grid with the new rectangle placed on it (we do not modify the original grid):

In [16]:
def place_rectangle_at(rect, grid, pos):
    """Place the rectangle of size (w, h) onto grid at position (x0, y0).
    Return a new grid, or None if the rectangle cannot be placed."""
    (w, h)   = rect
    (x0, y0) = pos
    newgrid  = [row.copy() for row in grid] 
    for x in range(x0, x0+w):
        for y in range(y0, y0+h):
            if y >= len(grid) or x >= len(grid[y]) or newgrid[y][x] is not empty:
                return None   
            newgrid[y][x] = rect
    return newgrid

In [17]:
# Place a 3 x 4 rectangle on a 5 x 5 square grid, 2 cells over and 1 down:
place_rectangle_at((3, 4), Square(5), (2, 1))

[[(0, 0), (0, 0), (0, 0), (0, 0), (0, 0)],
 [(0, 0), (0, 0), (3, 4), (3, 4), (3, 4)],
 [(0, 0), (0, 0), (3, 4), (3, 4), (3, 4)],
 [(0, 0), (0, 0), (3, 4), (3, 4), (3, 4)],
 [(0, 0), (0, 0), (3, 4), (3, 4), (3, 4)]]

In [18]:
# Place two rectangles on a grid
grid0 = Square(5)
grid1 = place_rectangle_at((3, 4), grid0, (2, 1))
grid2 = place_rectangle_at((2, 5), grid1, (0, 0))
grid2

[[(2, 5), (2, 5), (0, 0), (0, 0), (0, 0)],
 [(2, 5), (2, 5), (3, 4), (3, 4), (3, 4)],
 [(2, 5), (2, 5), (3, 4), (3, 4), (3, 4)],
 [(2, 5), (2, 5), (3, 4), (3, 4), (3, 4)],
 [(2, 5), (2, 5), (3, 4), (3, 4), (3, 4)]]

In [19]:
# Place a third rectangle that just barely fits
place_rectangle_at((3, 1), grid2, (2, 0))

[[(2, 5), (2, 5), (3, 1), (3, 1), (3, 1)],
 [(2, 5), (2, 5), (3, 4), (3, 4), (3, 4)],
 [(2, 5), (2, 5), (3, 4), (3, 4), (3, 4)],
 [(2, 5), (2, 5), (3, 4), (3, 4), (3, 4)],
 [(2, 5), (2, 5), (3, 4), (3, 4), (3, 4)]]

In [20]:
# Alternatively, try to place a third rectangle that does not fit
print(place_rectangle_at((4, 1), grid2, (2, 0)))

None


Now we need a strategy for packing a set of rectangles onto a grid.  I know that many variants of [bin packing problems](http://en.wikipedia.org/wiki/Bin_packing_problem) are NP-hard, but we only have 5 rectangles, so it should be easy: just exhaustively try each rectangle in each possible position, and in both possible orientations (horizontal and vertical). But placing rectangles is commutative, so we can do this two ways:

> Way 1: Considering the *rectangles* in a fixed order, try every possible *position* for each *rectangle*.

or

> Way 2: Considering the *positions* in a fixed order, try every possible *rectangle* for each *position*.

In Way 1, we could pre-sort the rectangles (say, biggest first).  Then we try to put the biggest rectangle in all possible positions on the grid, and for each position that fits, try putting the second biggest rectangle in all remaining positions, and so on.  As a rough estimate, assume there are on average about 10 ways to place a rectangle.  Then this way will look at about 10<sup>5</sup> = 100,000 combinations.

In Way 2, we consider the positions in some fixed order; say top-to-bottom, left-to right.  Take the first empty position (say, the upper left corner).  Try putting each of the rectangles there, and for each one that fits, try all possible rectangles in the next empty position, and so on. There are only 5! permutations of rectangles, and each rectangle can go either horizontally or vertically, so we would have to consider 5! &times; 2<sup>5</sup> = 3840 combinations.  Since 3840 &lt; 100,000, I'll go with Way 2.  Here is a more precise description:

> Way 2: To `pack` a set of rectangles onto a grid, find the first empty cell on the grid.  Try in turn all possible placements of any rectangle (in either orientation) at that position. For each one that fits, try to `pack` the remaining rectangles, and return the resulting grid if one of these packings succeeds.  

In [21]:
def pack(rectangles, grid):
    """Find a way to pack all rectangles onto grid and return the packed grid,
    or return None if not possible."""
    if not rectangles:
        return grid 
    pos = first_empty_cell(grid)
    if grid and pos:
        for (rectangles2, grid2) in rectangle_placements(rectangles, grid, pos):
            solution = pack(rectangles2, grid2)
            if solution:
                return solution

def rectangle_placements(rectangles, grid, pos):
    "Yield all (rectangles2, grid2) pairs that are the result of placing any rectangle at pos on grid."
    for (w, h) in rectangles:
        for rect in [(w, h), (h, w)]:
            grid2 = place_rectangle_at(rect, grid, pos)
            if grid2: 
                yield rectangles - {(w, h)}, grid2       
                
def first_empty_cell(grid):
    "The uppermost, leftmost empty cell."
    for (y, row) in enumerate(grid):
        for (x, cell) in enumerate(row):
            if cell is empty:
                return (x, y)

Let's try a simple example that I know will fit:

In [22]:
pack({(1, 3), (2, 5), (3, 4)}, Square(5))

[[(2, 5), (2, 5), (3, 1), (3, 1), (3, 1)],
 [(2, 5), (2, 5), (3, 4), (3, 4), (3, 4)],
 [(2, 5), (2, 5), (3, 4), (3, 4), (3, 4)],
 [(2, 5), (2, 5), (3, 4), (3, 4), (3, 4)],
 [(2, 5), (2, 5), (3, 4), (3, 4), (3, 4)]]

It works! But it is not pretty.

# Colored Rectangles

It would be nicer to have a graphical display of colored rectangles. I will define the function `show` which displays a grid as colored rectangles, by calling upon `html_table`, which formats any grid into HTML text.

In [23]:
from IPython.display import HTML, display, clear_output

def show(grid):
    "Display a colored HTML representation of this grid, unless it is None."
    if grid:
        display(html_table(grid, colored_cell))
    
def html_table(grid, cell_function='<td>{}'.format):
    "Return an HTML <table>, where each cell's contents comes from calling cell_function(grid[y][x])"
    return HTML('<table>{}</table>'
                .format(cat('\n<tr>' + cat(map(cell_function, row)) 
                        for row in grid)))

def colored_cell(rect): 
    x, y = sorted(rect)
    return '<td style="background-color:{}">{}{}'.format(colors[x], x%10, y%10)

colors = 'lightgrey yellow plum chartreuse cyan coral red olive slateblue lightgrey wheat'.split()

cat = ''.join

In [24]:
# For example:
show(pack({(1, 3), (2, 5), (3, 4)}, Square(5)))

That's much nicer to look at! Now we can show all the rectangles that are packable into a square:

In [25]:
def pack_square(rectangles):
    "Pack rectangles into a square of appropriate size, if possible."
    A = total_area(rectangles)
    if A in perfect_squares:
        return pack(rectangles, Square(int(A ** 0.5)))

for s in all_sets:
    show(pack_square(s))

So there are 4 sets of rectangles that work.

# Animated Colored Rectangles

It is gratifying to see the final results, and have the computation be so fast, but I'd like to get a better understanding of the process of finding the results.  I can try to visualize the process by *animating* the search that `pack` does: every time we successfully call `place_rectangle_at`, I can show the new grid, and then pause for a short period.  I accomplish this by modifying `pack` to take an optional argument, `animation`, which when false gives the same behavior as before, but when it is a number, pauses for that number of seconds in between displaying each step of the packing.

In [26]:
import time

def pack(rectangles, grid, animation=False): 
    """Find a way to pack all rectangles onto grid and return the packed grid,
    or return None if not possible. 
    Pause `animation` seconds between displaying each rectangle placement if `animation` is not false."""
    if animation: 
        clear_output()
        show(grid)
        time.sleep(animation) 
    if not rectangles:
        return grid 
    pos = first_empty_cell(grid)
    if grid and pos:
        for (rectangles2, grid2) in rectangle_placements(rectangles, grid, pos): 
            solution = pack(rectangles2, grid2, animation)
            if solution:
                return solution

If you are running this in a live IPython notebook (not in nbviewer), you can see for yourself by re-running this cell:

In [27]:
pack({(1, 2), (3, 7), (4, 6), (5, 10), (8, 9)}, Square(13), 1);

# Now For Something Completely Different: Cryptarithmetic

On April 28, Gary Antonik had another [Numberplay column](http://wordplay.blogs.nytimes.com/2014/04/28/num/) that quotes my friend Bill Gosper. (Gosper often presents more advanced puzzles in the [math-fun](http://mailman.xmission.com/cgi-bin/mailman/listinfo/math-fun) mailing list.)  This puzzle was:

>For the expression  N U M + B E R = P L A Y,

> Which distinct numerals (each different) can be substituted for letters to make a valid expression?

> How many solutions are there?

I tackled this type of problem (known as a [cryptarithmetic](http://mathworld.wolfram.com/Cryptarithmetic.html) or [alphametic](http://mathworld.wolfram.com/Alphametic.html) problem) in my Udacity class [CS 212](https://www.udacity.com/wiki/cs212/unit-2#cryptarithmetic). My approach was simple: try all permutations of digits replacing letters (that should be quick and easy&mdash;there can be at most 10 factorial or 3.6 million permutations), then for each one, use Python's `eval` function to see if the resulting string is a true expression.  The basic idea is simple, but there are complications to worry about:

1. Math uses `=` and Python uses `==` for equality; we'll allow either one.
2. The more detailed rules for these problems say that a number with a leading zero (like `012`) is illegal (but `0` itself is ok).
3. By default, Python thinks that 3/2 is 1, because integer division rounds down.  But we can import "proper" division from the future.
4. If we try to eval an expression like 1/0, Python raises an error; we'll have to handle it.

In [28]:
from __future__ import division
import itertools
import re

def solve_all(formula):
    """Given a formula like 'NUM + BER = PLAY', fill in digits to solve it.
    Input formula is a string; generate each digit-filled-in string."""
    formula = formula.replace(' = ', ' == ')
    for exp in replace_letters(formula):
        if valid(exp):
            yield exp

def replace_letters(formula):
    "Generate all possible replacements of letters with digits in formula."
    letters = ''.join(set(re.findall('[A-Z]', formula)))
    for digits in itertools.permutations('1234567890', len(letters)):
        trans = str.maketrans(letters, ''.join(digits))
        yield formula.translate(trans)

def valid(exp):
    "Expression is valid iff it has no numbers with leading zero, and evals true."
    try:
        return not leading_zero(exp) and eval(exp)
    except ArithmeticError:
        return False
    
leading_zero = re.compile(r'\b0[0-9]').search

In [29]:
next(solve_all('NUM + BER = PLAY'))

'857 + 349 == 1206'

In [30]:
%time len(list(solve_all('NUM + BER = PLAY')))

CPU times: user 29 s, sys: 42.2 ms, total: 29.1 s
Wall time: 29.1 s


96

So there are 96 solutions, but it is a bit slow to find them all. For each of the permutations of digits, we substitute the digits for letters, and then call  `eval(exp)` on the resulting string.  The majority of time is spent inside `eval`.  But the expression is basically the same each time; we could save a lot of work if we eval the expression just once, yielding a Python function, and then call that function once for each permutation of digits.  In other words, we want to take an expression such as

    "NUM + BER == PLAY"
    
and transform it into the Python function


    lambda A,B,E,M,L,N,P,R,U,Y: (100*N+10*U+M) + (100*B+10*E+R) == (1000*P+100*L+10*A+Y)
    
Actually that's not quite right.  The rules are that the first letter of each multi-letter word (the "N" in "NEW"; the "B" in "BER"; and the "P" in "PLAY") cannot be a zero. So the function should be:

    lambda A,B,E,M,L,N,P,R,U,Y: N and B and P and (100*N+10*U+M) + (100*B+10*E+R) == (1000*P+100*L+10*A+Y)

In my Udacity class I show a solution to this problem; you can [go there](https://www.udacity.com/wiki/cs212/unit-2#rethinking-eval) or see it here. Here is the code to compile a formula (or expression) into a Python function: 

In [31]:
def compile_formula(formula, verbose=False):
    """Compile formula into a function.   Also return letters found, as a str,
    in same order as parms of function. For example, 'YOU == ME**2' returns
    (lambda Y,M,E,U,O: Y and M and ((100*Y+10*O+U) == (10*M+E)**2)), 'YMEUO'"""
    formula      = formula.replace(' = ', ' == ')
    letters      = ''.join(set(re.findall('[A-Z]', formula)))
    firstletters = set(re.findall(r'\b([A-Z])[A-Z]', formula))
    body         = re.sub('[A-Z]+', compile_word, formula)
    if firstletters:
        body = '{} and ({})'.format(' and '.join(firstletters), body)
    fn = 'lambda {}: {}'.format(','.join(letters), body)
    if verbose: print(fn)
    return eval(fn), letters

def compile_word(matchobj):
    "Compile the word 'YOU' as '(100*Y + 10*O + U)'."
    word = matchobj.group()
    terms = reversed([mult(10**i, L) for (i, L) in enumerate(reversed(word))])
    return '(' + '+'.join(terms) + ')'

def mult(factor, var): return var if factor == 1 else str(factor) + '*' + var

In [32]:
compile_formula("YOU == ME**2", verbose=True)

lambda Y,M,E,O,U: Y and M and ((100*Y+10*O+U) == (10*M+E)**2)


(<function __main__.<lambda>>, 'YMEOU')

In [33]:
compile_formula("NUM + BER == PLAY", verbose=True)

lambda L,A,Y,P,M,E,R,U,B,N: P and B and N and ((100*N+10*U+M) + (100*B+10*E+R) == (1000*P+100*L+10*A+Y))


(<function __main__.<lambda>>, 'LAYPMERUBN')

In [34]:
def faster_solve_all(formula):
    """Given a formula like 'ODD + ODD == EVEN', fill in digits to solve it.
    Input formula is a string; output is a digit-filled-in string or None.
    Capital letters are variables. This version precompiles the formula; only one eval per formula."""
    fn, letters = compile_formula(formula)
    for digits in itertools.permutations((1,2,3,4,5,6,7,8,9,0), len(letters)):
        try:
            if fn(*digits):
                yield replace_all(formula, letters, digits)
        except ZeroDivisionError: 
            pass
        
def replace_all(text, olds, news):
    "Replace each occurrence of each old in text with the corresponding new."
    # E.g. replace_all('A + B', ['A', 'B'], [1, 2]) == '1 + 2'
    for (old, new) in zip(olds, news):
        text = text.replace(str(old), str(new))
    return text

In [35]:
next(faster_solve_all('NUM + BER = PLAY'))

'857 + 349 = 1206'

In [36]:
%time len(list(faster_solve_all('NUM + BER = PLAY')))

CPU times: user 1.75 s, sys: 3.29 ms, total: 1.76 s
Wall time: 1.76 s


96

We get the same answer, but 15 times faster.